arroyo-controller/src/job_controller/mod.rs

```diff
@@ -455,15 +455,14 @@ impl JobController {
             }
         }
 
-        if let Some(new_epoch) = self.model.compaction_needed() {
-            if self.compacting_task.is_none() && self.model.checkpoint_state.is_none() {
-                self.compacting_task = Some(self.start_compaction(new_epoch));
-            }
-        }
-
         // check on checkpointing
         if self.model.checkpoint_state.is_some() {
             self.model.finish_checkpoint_if_done(&self.pool).await?;
+            if let Some(new_epoch) = self.model.compaction_needed() {
```

Comment:
Yes, because we don't wait for a "state loaded" type signal there was a race condition where the files a subtask needed could get compacted away. By deferring compaction we know we'll be far enough ahead on the running pipeline.

Reply:
Ok, that seems like a bit of a hack around the actual problem but it's ok for now

Reply:
Actually, looking again I think that this is no longer a problem, since we now rely on the invariant that GlobalTables are fully written to every checkpoint, which is currently true in all cases.